### DISCUSSION TITLE
Information about kernel thread states - runqueue and blocked

### DISCUSSION THREAD

**Discussion by mnellemann** [oshi/oshi#2731] [Ideas]
I'm looking for a way to monitor running/run-queue and blocked (i/o waiting) threads and would like to hear if it makes sense to add this to oshi.

On Linux this can be viewed with the 'vmstat' command or by reading '/proc/stat'. On AIX and probably most other UNIX-like operating systems you also have the 'vmstat' command.

Br,
Mark

**Comment by dbwiddis** [comment]
Hey @mnellemann,

Is this at the OS level or for a particular process or thread?  I'm open to discuss adding some more available metrics, but generally want a consistent cross-platform interpretation of those metrics rather than something unique to any given OS.

OSHI does have a util/driver class, `ProcessStat`, which returns a map of all the `/proc/stat` values.  If you're looking for Linux-only information, the `getPidStats(int pid)` method might suffice for you.

We also query vmstat for memory statistics and could perhaps refactor to provide more information from those methods if needed.

**Comment by mnellemann** [comment]
Hi @dbwiddis,

I understand why you don't want unique OS metrics and will add it to my [project](https://github.com/mnellemann/sysmon) if it does not fit right with OSHI. I believe it's relevant for monitoring performance, but might not be general useful.

The `/proc/stat` on my Linux returns the `procs_running` and `procs_blocked` (near the end), but does not seem to be present in `driver/linux/proc/ProcessStat`. Maybe this was added in more recent kernels?

On AIX it's available through the `libperfstat.h`, `runque` in `perfstat_cpu_t`, or through the `vmstat` command output as mentioned.

**Comment by mnellemann** [comment]
I forgot to add: this is OS level.

The `running` queue should ideally be less than your number of logical processors or it could be an indication of not enough processor resources available for your workload.

The `blocked` queue could indicate bottlenecks in network or storage and cant be addressed by adding more processing resources.

Both to be monitored over a longer period of time to indicate any useful trends.

**Comment by dbwiddis** [comment]
Ah, I mentioned `ProcessStat` thinking it was a per-process thing.  We do have the `CpuStat` class that collects the cpu ticks via `/proc/stat` so we could collect those extra metrics there (which we don't currently do).

This is the same stat that contributes to load average, isn't it?  Likely some windows performance counters that report that as well.

Seems like a good idea for a feature!
